### DISCUSSION TITLE
Recommended capture device for Raspberry Pi?

### DISCUSSION THREAD

**Discussion by sofakng** [awawa-dev/HyperHDR#105] [General]
Can anybody recommend a capture device for the Raspberry Pi?

I don't think I need 4K HDR but it would be nice for something that supports this to future proof my setup.

My main concern is lag/latency so I'd like something as fast as possible (?)

**Comment by awawa-dev** [comment]
Best video quality, best internal splitter & fastest (they claim 50ms): Ezcap 320. Cons: Poor design/quality control, problems with a firmwares, mine won't work in USB3.0 but in USB2.0 with Rpi4.  
Best budget solution without an internal splitter, latency is good around 200-300ms depending on the video mode: MS2109 clones for around $10. Problematic free.  
  
So it's a choice between $10 and $100 in my opinion. Didn't find anything interesting between.

**Reply by sofakng** [reply]
Thanks so much for the information.  I should have mentioned that I don't need an HDMI splitter.  I'm either going to use my Denon X4300H receiver (it has 2x HDMI output) or an HDfury Vertex/Vertex2.

It sounds like the MS2109 clones might be a good starting point.  Are there any that do support USB 3.0 and YUV?  (and I don't need a loop/output like mentioned)

**Reply by awawa-dev** [reply]
First models supported MJPEG (1080p30) and YUV (although at lower framerate and resolution). They uses USB2.0. Then they started to experimenting and there are clones of that clones using 'USB3.0': they are rated for 1080p60 but only MJPEG..maybe something has changed. You must check specification of particular model.

**Comment by sofakng** [comment]
Got it.  One last question ... what's the full-chain latency like on the
MS2109 and Raspberry Pi 4?

I'm watching some Hyperion videos on YouTube and it looks like a full
second or so of latency instead of instant reaction?

**Reply by awawa-dev** [reply]
>Got it.  One last question ... what's the full-chain latency like on the MS2109 and Raspberry Pi 4?

Frame processing for 360p is <= 6ms. The only way to break things is to screw up 'smoothing settings' or to choose some very slow LED driver like Adalight (115200 or even 500 000bps)..or to connect to the LED driver over poor quality Wifi. Don't know what is your plan but if the LED strip is not SPI driven like APA102 then consider seriously cheap ESP8266 wemos D1 mini with CH340G or CP2104 or ESP32 also with CH340G or CP2104 as LED driver (just make sure that the seller can provide it in that specification because there are a lot of counterfeited items on the market).


> I'm watching some Hyperion videos on YouTube and it looks like a full
second or so of latency instead of instant reaction?

Completely without surprise... I've done research last year: here are the results https://hyperhdr.blogspot.com/2020/11/hyperhdr-prepare-for-building-buying_14.html . Don't know what is the current situation because they hide statistics from the user quite well. The main goal behind HyperHDR is to provide low latency system. In fact with disabled smoothing using Ezcap 320 and Rpi4 I have reaction from the LED strip dozen milliseconds before TV starts changing the image (some of that success comes probably due to fact that Ezcap is receiving the image first and TV also has some latency in the 'movie' mode).

**Comment by sofakng** [comment]
I was planning to use an ESP32 instead of ESP8266 unless there is a reason
to stick with the ESP8266?

Thanks so much for your articles.  It looks like you recommend the APA102
LEDs but that the non-counterfeit versions are difficult to find.  What you
instead suggest SK6812 RGBW (from BTF-Lighting perhaps?)

**Reply by awawa-dev** [reply]
ESP32 is also OK as long as it has a real ch340G or CP2104 onboard (for fast 2Mb USB serial port transfer in case of troubles with WiFi)  
    
I have many LED strips from BTF-Lighting . I was happy with them till the last purchase: sk6812 RGBW cold white started to fail after few days (one led after another, different segments). Probably bad luck. Next I purchased 96led/meter sk6812 but from other supplier. Personally I found RGBW cold much better than neutral. Cold white is perfectly balanced, neutral white is a bit yellow. Depends on your taste ...

**Comment by sofakng** [comment]
Oh, OK ... I see what you're saying.  I thought the CH340G / CP2104 were on
the LED strip for some reason.  It looks like I have a bit of reading to do.

Also, I thought an ESP32/ESP8266 were required but I think I'm wrong?  I'm
going to be using a Raspberry Pi 4 directly connected to the SK6812 LEDs
(if possible, which I think so?).  I was just confused because of your web
page recommendation of APA102.

I've been trying to research and it looks like the SK6812 also uses SPI so
it should be equally as fast?

**Reply by awawa-dev** [reply]
No, Sk6812 is difficult one (non-SPI) same as WS2812b...the success rate with the direct connection to the Rpi is even much, much lower. You can find a lot of stories about that problem here. ESP32/ESP8266 is a must for that LED strip. APA102 (or fake: APA102C/SK9812) is completely different story (SPI, you don't need ESP).

**Reply by sofakng** [reply]
OK - I was avoiding the APA102 since you didn't recommend it due to the counterfeits and difficulty finding a reputable vendor.  It sounds like I'll need to go the RPi -> ESP32 (w/CH430G or CP2104).  Hopefully the latency will still be very low.

I wonder if you can plug the ESP32 into the RPi via USB to eliminate wireless latency?  I've also found a few ESP32 boards with ethernet that could help as well.  (...or maybe even connect the ESP32 to the RPi using I2C or SPI or something?)

**Reply by awawa-dev** [reply]
Counterfeited APA102 has a tendency to die faster and their specification is inferior compared to the genuine APA102. But the biggest advantage of sk6812 is RGBW: the quality of the ambient lighting is a way better then RGB LED strip. And there is no alternative for that solution (well, there is in fact HD108 RGBW but the price is out of this world)

> I wonder if you can plug the ESP32 into the RPi via USB to eliminate wireless latency  

That's the point: you plug the ESP32 into the USB and have no problems with Wifi ;) Communication is over USB, you don't need an Ethernet or wifi.  
  
Lets assume that you have 300led ->frame's size is ~900bytes that's 7200bits -> we need to transfer it using USB @2Mb/sec -> in perfect world it takes about 0,0036 second to transfer a frame to the LED driver.  
If it is still to much you can always use my new [HyperSPI](https://github.com/awawa-dev/HyperSPI) project and have 25Mb speed 😉 But it's more complicated to wire all the things together than adalight@2Mb.
